### PR TITLE
BGDIINF_SB-1977: sphinxsearch as container with EFS support

### PR DESCRIPTION
**Pull request by ltclm**
- [x] index data is persisted on efs
- [x] index can be shared between multiple containers
- [x] service endpoint -> default CMD
- [x] makefile endpoints
  - [x] dockerbuild (image tagging as proposed here https://github.com/geoadmin/doc-guidelines/pull/80/files )
  - [x] dockerrun
  - [x] dockerrestart
  - [x] checkconfig
  - [x] index
  - [x] template
  - [x] move-template
- [x] git pre-commit hook ( check-config ) 
- [x] terraform pr
- [x] ECR push
- [x] E2E and integration tests (search test on all layers) will be done here https://github.com/geoadmin/service-search-wsgi
- [x] Add the AWS CodeBuild in the required status for the master and develop-* branch protection (only after merge of this)

### CODE REVIEW

File: Makefile
Comment (ltshb) [inline Makefile:113]: Maybe it would be better to have the two in one target checking both ? So when working you don't have to bother running both when you change both python and bash code, simply run one.

File: Makefile
Comment (ltshb) [inline Makefile:249]: Mixing spaces and tab.

File: Pipfile
Comment (ltshb) [inline Pipfile:9]: Probably not needed if not a flask application

File: dev.env
Comment (ltshb) [inline dev.env:4]: Although those credential are not really real one (easily guessable) as a best practice we should never put credential in a git repository in plain text. Might be a good idea to use `gopass` here ? Not sure though if it will make the workflow for sphinxsearch too complicated if using gopass ?

File: scripts/check-config-local.sh
Comment (ltshb) [inline scripts/check-config-local.sh:18]: Styling question to be consistent. I would prefer an `if  [...]` here it would make it more readable and avoid such long line.

File: scripts/docker-cmd.sh
Comment (ltshb) [inline scripts/docker-cmd.sh:28]: I would not mix coding style and use also 
```bash
if [[ -z "${index}" ]]; then
  continue
fi
```
Also use `"` in both `if` with `${index}` and be consistent with `${index}` or `$index`

File: scripts/docker-cmd.sh
Comment (ltshb) [inline scripts/docker-cmd.sh:35]: Same as above

File: scripts/docker-cmd.sh
Comment (ltshb) [inline scripts/docker-cmd.sh:36]: Styling consistency with if, why not using `if [[ ... ]]` here ? should it not work with `eval` ?

File: scripts/docker-entry.sh
Comment (ltshb) [inline scripts/docker-entry.sh:10]: Could this not be done in dockerfile during the build instead of during each docker start ?

File: scripts/pg2sphinx_trigger.py
Comment (ltshb) [inline scripts/pg2sphinx_trigger.py:14]: psycopg2 should be part of the Pipfile and then you won't have this pylint error

File: scripts/pg2sphinx_trigger.py
Comment (ltshb) [inline scripts/pg2sphinx_trigger.py:12]: Why not yet use `argparse` to avoid this warning. This module might be soon not be anymore supported and we will need then to do the changes anyway.

File: scripts/pg2sphinx_trigger.py
Comment (ltshb) [inline scripts/pg2sphinx_trigger.py:23]: Might be a good idea to use f string here, more readable and also more modern as the old style string formating.

File: dev.env
Comment (boecklic) [reply dev.env:4]: These env variables are part of the infra definition and should live in the vhost setup for service search in `infra-vhost` repository

File: int.env
Comment (boecklic) [inline int.env:6]: dito, see above

File: prod.env
Comment (boecklic) [inline prod.env:6]: dito

File: scripts/docker-cron.sh
Comment (boecklic) [inline scripts/docker-cron.sh:1]: detail: the script does index update/index rotation, so a name like `index-rotation.sh` would be more descriptive than `docker-cron.sh`, even if the script is executed from a cron job

**Review by boecklic — COMMENTED** [review]
Nice work!! 

I have one general question: what is this container logging? And in which format? If just the scripts produce log output, I'd suggest to check if the output could be valid json (similar in format/fields etc. to all other services) that we could feed this directly into kibana without extra processing needed. If sphinx is logging as well we would need to check with probably Sam how we can format / parse the logs to get them into kibana...

File: Dockerfile
Comment (ltclm) [reply Dockerfile:21]: thank you, yes i have tried to re-group the RUN commands

File: Dockerfile
Comment (ltclm) [reply Dockerfile:31]: i have replaced the two run execution with 
```bash
install -o geodata -g geodata -d /var/lib/sphinxsearch/data/index/ && \
install -o geodata -p geodata -d /var/lib/sphinxsearch/data/index_efs/ && \
```
this created the folder hierarchy and sets the ownership

File: Dockerfile
Comment (ltclm) [reply Dockerfile:28]: done

File: Dockerfile
Comment (ltclm) [reply Dockerfile:36]: they can belong to root.
Im running ``exec gosu geodata "$@"`` in the entrypoint now.
this makes the whole installation easier.

File: Dockerfile
Comment (ltclm) [reply Dockerfile:33]: yes they are created with apt-get install. i have regrouped now everything in the same run command

File: scripts/docker-cmd.sh
Comment (ltclm) [reply scripts/docker-cmd.sh:36]: yes, good point, i will change this.

File: scripts/docker-entry.sh
Comment (ltclm) [reply scripts/docker-entry.sh:10]: the reason for having this here is that we can use the same image for dev, int and prod staging. envsubst will replace the db suffix etc. when run the image/container. otherwise we would have to build one image per staging.

File: scripts/pg2sphinx_trigger.py
Comment (ltclm) [reply scripts/pg2sphinx_trigger.py:14]: it is only running inside the container. do we use virtualenv inside the container too?

Comment (ltshb) [reply scripts/pg2sphinx_trigger.py:14]: Ok I see you use the apt `python3-psycopg2` in Dockerfile. Then for local development with IDE like vscode and co I would suggest to add psycopg2 to the Pipfile dev dependencies and remove the pylint error. A comment in pipfile why it is in dev dependencies might be welcome.

File: scripts/docker-entry.sh
Comment (ltshb) [reply scripts/docker-entry.sh:10]: OK sorry I missed the envsubst part above. That make now total sense.

File: conf/uvek.conf.part
Comment (hansmannj) [inline conf/uvek.conf.part:1658]: There's one `_dev` left that is not replaced by `${DBSTAGING}` in line 1676, is this by intention or was it forgotton?

File: conf/vbs.conf.part
Comment (hansmannj) [inline conf/vbs.conf.part:132]: One `_dev` left in line 151 that hasn't been replaced by `${DBSTAGING}`. Intentionally, or has it been hiding? :wink:

File: scripts/pg2sphinx_trigger.py
Comment (hansmannj) [reply scripts/pg2sphinx_trigger.py:23]: Yes, and also would avoid rising `consider-using-f-string` with current pylint versions: https://pylint.pycqa.org/en/latest/whatsnew/2.11.html#new-checkers

File: scripts/pg2sphinx_trigger.py
Comment (hansmannj) [inline scripts/pg2sphinx_trigger.py:184]: Probably leave it `TODO` and adapt .pylintrc accordingly in order to "mute" this warning?

File: scripts/pg2sphinx_trigger.py
Comment (hansmannj) [inline scripts/pg2sphinx_trigger.py:229]: ditto

File: test/sphinxapi.py
Comment (hansmannj) [inline test/sphinxapi.py:641]: I get a `"SPH_FILTER_STRING" is not defined` here locally

File: test/sphinxapi.py
Comment (hansmannj) [inline test/sphinxapi.py:645]: And likewise a "SPH_FILTER_STRING_LIST" is not defined here, locally.

File: Dockerfile
Comment (boecklic) [reply Dockerfile:1]: Instead of using `debian:buster-slim` as base you could use `python:3.10.1-(slim-)buster` for hassle-free python version management. But using the default version of `buster-slim` is also ok

File: scripts/pg2sphinx_trigger.py
Comment (ltclm) [reply scripts/pg2sphinx_trigger.py]: i have migrated the project to python3.9.9 and im using fstrings now

**Comment by ltclm** [comment]
@boecklic @ltshb i have to use python:3.9.9-slim-buster now because debian-bullseye-slim has no sphinxsearch suppport.

File: int.env
Comment (ltclm) [reply int.env:6]: @ltshb @boecklic and how can we use them without the infra-vhost project / docker-compose file. 
p.e. the search index maintenance (index updates, etc)  will not be executed with the infra-vhost setup, this will be executed as ``DB=stopo_dev.public.ch make pg2sphinx`` on the data development host.

File: conf/uvek.conf.part
Comment (ltclm) [reply conf/uvek.conf.part:1658]: 😨 forgotten
thank you!

File: conf/vbs.conf.part
Comment (ltclm) [reply conf/vbs.conf.part:132]: done

**Comment by ltclm** [comment]
> Nice work!!
> 
> I have one general question: what is this container logging? And in which format? If just the scripts produce log output, I'd suggest to check if the output could be valid json (similar in format/fields etc. to all other services) that we could feed this directly into kibana without extra processing needed. If sphinx is logging as well we would need to check with probably Sam how we can format / parse the logs to get them into kibana...

@boecklic 
sphinxsearch is writing  a searchd log and a query log:
```bash
# /var/log/sphinxsearch/searchd.log
[Thu Dec 16 20:12:19.034 2021] [2093] caught SIGHUP (seamless=1, in queue=1)
[Thu Dec 16 20:12:19.073 2021] [2094] rotating index 'layers_rm': started

# /var/log/sphinxsearch/query.log
[Thu Dec 16 17:06:37.915 2021] 0.001 sec 0.001 sec [ext/1/ext 0 (0,50)] [swisssearch_fuzzy] @detail "meteoschweiz.klimanormwerte" | @detail "^meteoschweiz.klimanormwerte" | @detail "meteoschweiz.klimanormwerte$" | @detail "^meteoschweiz.klimanormwerte$" | @detail "meteoschweiz.klimanormwerte"~5 | @detail "meteoschweiz.klimanormwerte*" | @detail "^meteoschweiz.klimanormwerte*" | @detail "meteoschweiz.klimanormwerte*"~5 | @detail "*meteoschweiz.klimanormwerte*" | @detail "^*meteoschweiz.klimanormwerte*" | @detail "*meteoschweiz.klimanormwerte*"~5
```

the searchd log is only showing searchd service logs (restarts, index rotation, etc.). index requests are in the query log.

**Comment by ltshb** [comment]
@ltclm once this PR is merged, could you add the `AWS CodeBuild` in the required status for the `master` and `develop-*` branch protection.

**Comment by ltclm** [comment]
> @ltclm once this PR is merged, could you add the `AWS CodeBuild` in the required status for the `master` and `develop-*` branch protection.

sure, but it may take a while before we can merge this PR. has to be done together with or after the migration of geodata pipeline/automata to Frankfurt.

File: Makefile
Comment (ltshb) [reply Makefile:257]: Oops the indentation should be also TAB here like the rest

File: Pipfile
Comment (ltshb) [reply Pipfile:14]: pylint-flask should be removed

File: int.env
Comment (ltshb) [reply int.env:6]: @ltclm as I don't have another proposal for this use case it is fine for me to leave these infra settings in the repo, however as a best practice I would avoid setting credential in plain text here (even though those are not real one :-)) and use gopass with summon. It is also better for maintenance if we in future change this credential we only have to change it in gopass and don't have to touch each repo that use them.

**Review by ltshb — COMMENTED** [review]
still some minor issues to fix, see comment above. But otherwise thanks for the great work.

**Comment by ltclm** [comment]
@rebert @ltshb @boecklic @hansmannj need new approval / review.
I have addressed most of your points.

**Review by ltshb — APPROVED** [review]
🥇  Great work many thanks

**Pull request merged**
